Datablob transactions

protolambda opened the pull request:
This PR tracks the Datablob Eth-Denver diff (building on `base`, v1.10.16)

- Proto: Implement data blob tx type (SSZ type, following the updated ethdenver spec)
- Proto: New "danksigner" to verify signature, get origin, and compute tx hash
- Proto: Update remote signer api
- Matt: EVM sharding fork definition
- Matt: DATAHASH opcode
- Matt: gas per blob
- Matt: intrinsic gas update with data hashes
- Matt: use danksigner in EVM
- Matt: data blob tx test
- Proto: transaction pool update blob-size check, placeholder check for blob <> kzg commit <> versioned hash check
- Proto: implement "txWrapData", optional addition to `*Transaction.inner`, and encode minimal (for blocks) vs wrapped (default, used in storage, tx pool, journal etc.).
- Proto: update JSON-RPC for tx wrapper data encoding/decoding
- Proto: update signer to compute commitments and versioned hashes, signer API caller only provides the blob data
- Proto: implement SSZ and JSON encoding for blobs and kzg commitments, for network wrapper
- Proto: update engine API block building
- Proto/George: update Go-KZG to marshal/unmarshal G1/G2 hex text form (also applies to json)
- George: Basic KZG functionality needed for danksharding
- George: KZG trusted setup data + loading + unittest
- George: blobVerification precompile + test vector (happy case)
- George: pointEvaluation precompile + test vector (happy case)
- Matt/Proto: engine API getBlob method (relatively hacky, merging with `ExecutionPayload` response may make more sense)
- Proto: don't reinject incomplete blob transactions into tx pool on reorg
- Proto: work on RLP block encoding, do not encode full blob txs with wrapped data
- Matt: intrinsic gas check in tx pool considers data blobs now
- George: isolate minimal KZG code for geth, avoid library import (BLS types left to do)
- Proto: update engine API with `getBlobsBundleV1` as [proposed](https://github.com/ethereum/execution-apis/pull/197) in execution APIs repo 
- Proto: update more signer api types to build blob txs with
- Proto: update `Blob`/`KZGCommitment`/`BLSFieldElement` types text encoding/decoding. (and `Blob` is array type now)
- George: `Parse()` methods on `Blobs`/`Blob`/`KZGCommitments`/`KZGCommitment`
- George: `kzg.VerifyBlobs(commitments, blobs)` for efficient batch verification.
- Proto: implement `kzg.BlobsBatch` to run same batch verification, but aggregate on the fly, to handle batching of many txs
- Proto: refactor tx wrap-data verification to support vatch-verification
- Proto: refactor tx pool tx-addition to be more batch-centric, deprecate adding single tx, split validation in stages, and add a blobs batch-verification stage.
- George: interop bugfixes (missing versioned hashes, use latest signer, etc.)

Consensus layer:
- Prysm (work in progress): see https://github.com/prysmaticlabs/prysm/compare/kiln...blobs

lightclient committed:
params/conf: add sharding fork defn

lightclient committed:
all: add data hashes to evm msg

lightclient committed:
core/vm: impl DATAHASH opcode

lightclient committed:
params: update gas per blob parameter to 120k

asn-d6 committed:
Import go-kzg to geth

asn-d6 committed:
Introduce dummy KZG trusted setup generator

asn-d6 committed:
Introduce KZG crypto module (that implements the funcs from the mini-danksharding EIP)

asn-d6 committed:
Introduce unittests for KZG and the trusted setup

protolambda committed:
core/types: data blob transaction type and signer

protolambda committed:
accounts,signer: blob tx signer support

lightclient committed:
core/vm: use correct gas price for DATAHASH

lightclient committed:
core/types/data_blob_tx: grab correct r,s values from signature

lightclient committed:
core/types/transaction_signing: use dank signer when needed

lightclient committed:
core/types/transaction: grab data hashes from blob txs for message

lightclient committed:
core: add test for data blob txs

asn-d6 committed:
Add the trusted setup json data as a go file

lightclient committed:
all: fix intrinsic gas calls in various tests

asn-d6 committed:
Introduce blobVerification precompile

protolambda committed:
Merge pull request #2 from lightclient/simple-blob

Implement `DATAHASH` opcode and sharding fork

asn-d6 committed:
Introduce blobVerification precompile test vector

protolambda committed:
work in progress, tx wrapper for blob tx pool / journal / block building / engine API

asn-d6 committed:
Introduce pointEvaluation precompile

asn-d6 committed:
Introduce pointEvaluation precompile test vector (happy case)

protolambda committed:
core/types: implement data blob / kzg and lists ssz encoding

protolambda committed:
eth/catalyst: GetBlobV1

protolambda committed:
go.mod,go.sum: update bls/kzg libs

protolambda committed:
go.mod,go.sum: update KZG lib

protolambda committed:
core/vm,crypto/kzg,tests: update data blob kzg to use From/To Compressed, and change snake_case to camelCase

protolambda committed:
core/types,crypto/kzg,signer: compute KZG commitment from raw blob data

asn-d6 committed:
Update test vectors for blobVerification/pointEvaluation

They broke when we changed blake2b to keccak in the versioned hash.

asn-d6 committed:
Remove go-kzg dependency from kzg crypto module

We still use go-kzg for tests and cmd/kzg_dummy_setup. To remove that dependency there we would need to add a bunch of
go-kzg code to go-ethereum or simplify the tests.

Next step is to also remove the `go-kzg/bls` dependency and just use the one included in geth.

protolambda committed:
core/types: blob kzgs list ssz vector->list fix, fix blob tx encoding/decoding

protolambda committed:
core/types: test dynamic fee and blob tx binary+json encoding

protolambda committed:
core: don't reinject incomplete transactions on reorg

protolambda committed:
core/types: encode/decode the block with transactions the minimal way

asn-d6 commented:
Please see https://github.com/protolambda/go-ethereum/pull/3 for an "incremental" simplification of the KZG stack.
The next step is also removing the `go-kzg/bls` requirement and using the geth bls, but that requires a bunch of functions to be added to it (e.g. lincomb)

asn-d6 commented on core/types/data_blob.go:
```diff
@@ -0,0 +1,332 @@
+	return kbls.NewG1().FromCompressed(p[:])
+}
+
+func (kzg KZGCommitment) ComputeVersionedHash() common.Hash {
```
Code dup with `crypto/kzg/kzg.go:KzgToVersionedHash`

protolambda replied:
Removed the duplicate in kzg package, updated the tests / usage. The versioned-hashes are an ethereum application detail, not really a KZG thing.

protolambda committed:
core: consider blobs count for intrinsic gas check

Co-authored-by: lightclient <14004106+lightclient@users.noreply.github.com>

protolambda committed:
Merge pull request #3 from asn-d6/prune-gokzg

Prune gokzg from crypto/kzg

asn-d6 committed:
Add functions to extract the crypto out of Blobs/KZGCommitment/BlobKzgs

asn-d6 committed:
Use optimized VerifyBlobs() function

asn-d6 committed:
Add unittest for VerifyBlobs() function

asn-d6 committed:
kzg: Rename CRS variables to match the EIP naming

asn-d6 committed:
kzg: Document kzg code further

asn-d6 commented on core/types/data_blob.go:
```diff
@@ -0,0 +1,332 @@
+			return fmt.Errorf("versioned hash %d supposedly %s but does not match computed %s", i, h, computed)
+		}
+	}
+	// TODO: george/dankrad: faster check if kzg commitment matches blob data, Dankrad: "Instead of executing
```
Hey hey. This is resolved by https://github.com/protolambda/go-ethereum/pull/4

protolambda committed:
eth/catalyst: implement GetBlobsBundleV1 - EIP-4844 engine api extension

terencechain commented:
Got the following when trying to `make` on `587a212`:
````
# github.com/ethereum/go-ethereum/signer/core/apitypes
signer/core/apitypes/types.go:162:10: undefined: types.NewTxWrapped
```

protolambda committed:
core/types: compute kzgs for blobs list

protolambda committed:
accounts,internal/ethapi: blob transaction api arguments

protolambda committed:
signer/core/apitypes: fix wrap-data NewTx option

protolambda commented:
@terencechain 5fa43da2fadbb4b84bfec260ae79abf62aef4c20 fixes the build issue, and I updated some more of the signer api to build blob transactions. It builds, but still experimental. The surface area to create/sign new transactions is huge, there is a lot to update.

protolambda committed:
core/types: fix marshal/unmarshal text Blob type, BLSFieldElement, KZGCommitment

protolambda committed:
core/types: clean up blob and commitment parse methods, use ptr receiver on array blob type

protolambda committed:
tests: blob test update, Parse blobs/commitments, fix mutation on copy

protolambda committed:
params,core/types: remove duplicate kzg version param

protolambda committed:
core/vm,crypto/kzg,tests: remove versioned-hash func from crypto, application-layer detail, update tests. And make kzg tests run precompiles.

protolambda committed:
tests: benchmark kzg funcs

protolambda committed:
crypto/kzg: batch verification for multiple txs

protolambda committed:
core/types: support blobs batch-verify of multiple txs

protolambda committed:
core: refactor TxPool.add and .addTxsLocked to batch-verify txs

asn-d6 commented on crypto/kzg/kzg.go:
```diff
@@ -37,6 +39,63 @@ func VerifyKzgProof(commitment *bls.G1Point, x *bls.Fr, y *bls.Fr, proof *bls.G1
+	var tmpFr bls.Fr
+	for i := 0; i < params.FieldElementsPerBlob; i++ {
+		bls.MulModFr(&tmpFr, &blob[i], randomScalar)
+		bls.AddModFr(&batch.aggregateBlob[i], &batch.aggregateBlob[i], &tmpFr)
```
Naming nitpick: `s/aggregateBlob/aggregateScalar`

It's not actually a single blob that is aggregated in every `aggregateBlob` scalar; it's actually all the blobs.

asn-d6 committed:
Update LatestSignerForChainID() to return a danksigner by default

asn-d6 committed:
Transaction inner size should be computed on-demand

We previously computed it in UnmarshalBinary() but in a wrong way (we considered the wrapped data as part of the inner
length)

asn-d6 committed:
Include versioned hashes when creating a transaction using the JSON-RPC API

asn-d6 commented:
A bunch of bug fixes found while developing [blobber](https://github.com/asn-d6/blobber) can be found here: https://github.com/protolambda/go-ethereum/pull/5

protolambda committed:
Merge pull request #5 from asn-d6/eip4844-tooling

Fix a bunch of bugs found during development of EIP-4844 tooling

protolambda committed:
remove blob verification precompile in favor of point evaluation precompile

protolambda committed:
fix lint
